### PR TITLE
`topk`: port to structured

### PR DESCRIPTION
**Pull request by eqy**
#55070

There are a few places where `const_cast` is used with utility functions shared with unstructured operators.
The RFC says that assigning to the `out` tensor doesn't work, but that seems to be what e.g., `_allocate_or_resize_output_with_indices` seems to do. Does assignment "work" when the tensors are not allocated?

### CODE REVIEW

**Comment by ezyang** [comment]
> The RFC says that assigning to the out tensor doesn't work, but that seems to be what e.g., _allocate_or_resize_output_with_indices seems to do. 

This function is wrong for doing that and should get fixed.

File: aten/src/ATen/native/cuda/TensorTopK.cu [hunk 1]
```diff
@@ -150,22 +150,21 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
...
   numDims = numDims == 0 ? 1 : numDims;
   TORCH_CHECK(numDims <= MAX_DIMS, "input tensor has too many dimensions");
-  TORCH_CHECK(dim >= 0 && dim < numDims, "dim not in range");
-
   int64_t sliceSize = self.dim() == 0 ? 1 : self.size(dim);
-  TORCH_CHECK(k >= 0 && k <= sliceSize, "k not in range for dimension");
+
   Tensor input = self.contiguous();
 
   // Build the output size, which is the dim being selected set to
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/TensorTopK.cu:170]: The logic here and below should be removed; structured kernels handle resizing for you

File: aten/src/ATen/native/Sorting.cpp [hunk 2]
```diff
@@ -706,40 +733,26 @@ std::tuple<Tensor, Tensor> kthvalue(
...
-    Tensor& indices) {
+    const Tensor& values,
+    const Tensor& indices) {
   int64_t dim = maybe_wrap_dim(dim_, self.dim(), /*wrap_scalar=*/true);
   TORCH_CHECK(
       k >= 0 && k <= (self.dim() > 0 ? self.size(dim) : 1),
       "selected index k out of range");
 
-  _allocate_or_resize_output_with_indices(values, indices, self, dim_, k);
+  _allocate_or_resize_output_with_indices(const_cast<Tensor&>(values), const_cast<Tensor&>(indices), self, dim_, k);
```
Comment (ezyang) [inline aten/src/ATen/native/Sorting.cpp:749]: Delete this entirely.

File: aten/src/ATen/native/Sorting.cpp [hunk 3]
```diff
@@ -706,40 +733,26 @@ std::tuple<Tensor, Tensor> kthvalue(
...
     bool largest,
     bool sorted,
-    Tensor& values,
-    Tensor& indices) {
+    const Tensor& values,
+    const Tensor& indices) {
   int64_t dim = maybe_wrap_dim(dim_, self.dim(), /*wrap_scalar=*/true);
   TORCH_CHECK(
       k >= 0 && k <= (self.dim() > 0 ? self.size(dim) : 1),
       "selected index k out of range");
```
Comment (ezyang) [inline aten/src/ATen/native/Sorting.cpp:747]: You can delete these checks because they are guaranteed to be satisfied by this point.

**Review by ezyang — CHANGES_REQUESTED** [review]
This is a good first attempt. Remember that every check you put in the meta function is guaranteed to hold by the time you do the impl function, so you can delete all checks in the impl function that are now redundant.

File: aten/src/ATen/native/cuda/TensorTopK.cu (hunk 1 above)
Comment (eqy) [reply aten/src/ATen/native/cuda/TensorTopK.cu:170]: Right, I think those can mostly be removed but I'm not sure about the `MAX_DIMS` check as that's not in Sorting.cpp where the Meta function is defined.

Comment (ezyang) [reply aten/src/ATen/native/cuda/TensorTopK.cu]: I agree with you that `MAX_DIMS` can stay in the kernel. It looks like a kernel implementation detail and the semantic meaning of the kernel is that it should support arbitrary numbers of dims.

**Review by ezyang — APPROVED** [review]
Thanks, nice work!

**Pull request closed without merging**